File: ts/models/skill.ts
import { Attribute } from './attribute'
import { PurchaseCost } from './purchase'
import { RangeType } from './range'
import { TargetMechanics, TargetType } from './target'
import { UpgradeCost } from './upgradeCost'

/**
 * Shows the minimum level required for the player to level up their skill to a certain level.
 */
export interface SkillLevelRequirement {
    // the skill level
    level: number,
    // the minimum player level required to level up the skill to this level
    minPlayerLevelRequired: number,
}

/**
 * Represents an enemy's skill.
 */
export interface EnemySkill {
    // the skill's attribute (i.e. characteristics)
    attributes: Attribute[],
    // the skill's additional description (on top of the attribute's description if needed)
    addDescription: string,
    // the skill's level mechanics for the attributes
    levelMechanics: EnemySkillLevelMechanics[],
}

/**
 * Represents a skill that deals damage (mainly)
 */
export interface DamagingSkill {
    // the skill's name
    name: string,
    // the skill's description
    description: string,
    // the skill's targetting mechanics
    targetMechanics: TargetMechanics,
    // the skill's range type
    rangeType: RangeType,
    // the skill's activation range
    range: number,
    // the cost to purchase this skill (if it is a purchasable skill)
    purchaseCost: PurchaseCost,
    // the skill's level mechanics
    levelMechanics: DamagingSkillLevelMechanics[],
}

/**
 * Represents a skill that does not deal damage (mainly)
 */
export interface NonDamagingSkill {
    // the skill's name
    name: string,
    // the skill's description
    description: string,
    // the skill's targetting mechanics
    targetMechanics: TargetMechanics,
    // the skill's range type
    rangeType: RangeType,
    // the skill's activation range
    range: number,
    // the cost to purchase this skill (if it is a purchasable skill)
    purchaseCost: PurchaseCost,
    // the skill's level mechanics
    levelMechanics: NonDamagingSkillLevelMechanics[],
}

/**
 * Represents a damaging skill's level mechanics (how it performs level by level)
 */
export interface DamagingSkillLevelMechanics {
    // the skill's level
    level: number,
    // the skill's base damage at this level
    baseDamage: number,
    // the skill's base cast time at this level
    baseSkillCastTime: number,
    // the skill's duration (0 if not status effect)
    duration: number,
    // the skill's base cooldown at this level
    cooldown: number,
    // the skill's attributes at this level
    attributes: DamagingSkillAttribute[],
    // the cost required to upgrade the skill to this level
    upgradeCost: UpgradeCost,
}

/**
 * Represents a non-damaging skill's level mechanics (how it performs level by level)
 */
export interface NonDamagingSkillLevelMechanics {
    // the skill's level
    level: number,
    // the skill's base cast time at this level
    baseSkillCastTime: number,
    // the skill's duration in seconds (only applies if there is a status effect with a duration), else 0
    duration: number,
    // the skill's base cooldown at this level
    cooldown: number,
    // the skill's attributes at this level
    attributes: NonDamagingSkillAttribute[],
    // the cost required to upgrade the skill to this level
    upgradeCost: UpgradeCost,
}

/**
 * Represents an attribute for a non-damaging skill.
 */
export interface NonDamagingSkillAttribute {
    // the skill's attribute
    attribute: Attribute,
    // the chance for the attribute to play out
    chance: number,
    // the skill's attribute modifier
    modifier: SkillModifier,
    // the radius of effect of the skill's attribute
    radius: number,
}

/**
 * Represents an attribute for a damaging skill.
 */
export interface DamagingSkillAttribute {
    // the skill's attribute
    attribute: Attribute,
    // the chance for the attribute to play out
    chance: number,
    // the skill's attribute modifier
    modifier: SkillModifier,
    // the radius of effect of the skill's attribute
    radius: number,
}

/**
 * Represents an enemy's skill data level by level.
 */
export interface EnemySkillLevelMechanics {
    // the skill level
    level: number,
    // the skill's duration in seconds (only applies if there is a status effect with a duration), else 0
    duration: number,
    // the skill's modifier
    modifier: SkillModifier,
    // the chance for the skill to play out
    chance: number,
    // the skill's radius of effect
    radius: number,
}

/**
 * Represents a skill's modifier.
 */
export interface SkillModifier {
    // the type of the modifier
    type: SkillModifierType,
    // the value of the modifier
    value: number,
}

/**
 * Lists all current skill modifier types.
 */
export enum SkillModifierType {
    Percentage = 'Percentage',
    Number = 'Number',
}

File: rust/models/skill.rs
use serde::{Serialize, Deserialize};

use super::{Attribute, TargetMechanics, RangeType};
/// Represents a skill instance and its data.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct Skill {
    /// the skill's name
    name: String,
    /// the skill's description
    description: String,
    /// the skill's type (damaging or non-damaging)
    skill_type: SkillType,
    /// the skill's targetting mechanics
    #[serde(rename = "targetMechanics")]
    target_mechanics: TargetMechanics,
    /// the skill's range type
    #[serde(rename = "rangeType")]
    range_type: RangeType,
    /// the skill's activation range
    #[serde(rename = "activationRange")]
    activation_range: f64,
    /// the skill's level
    level: u8,
    /// the skill's base damage (if damaging skill)
    #[serde(rename = "baseDamage")]
    base_damage: Option<f64>,
    /// the skill's cast time (in seconds)
    cast_time: f64,
    /// the skill's duration in seconds (0 if not status effect)
    duration: f64,
    /// the skill's cooldown (in seconds)
    cooldown: f64,
    /// the skill's attributes
    attributes: Vec<SkillAttribute>
}

/// Lists all available skill types.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum SkillType {
    #[serde(rename = "Damaging")]
    Damaging,
    #[serde(rename = "NonDamaging")]
    NonDamaging,
}

/// Represents a skill's attribute.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SkillAttribute {
    /// the skill's attribute
    attribute: Attribute,
    /// the chance for the attribute to play out (in a ratio from 0 to 1)
    chance: f64,
    /// the attribute's modifier
    modifier: SkillModifier,
    /// the attribute's radius of effect (in units)
    radius: f64,
}

/// Represents an enemy's skill data.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct EnemySkill {
    /// the skill's attribute (i.e. characteristics)
    attributes: Vec<Attribute>,
    /// an additional description to explain the skill (on top of the attribute's description if needed)
    #[serde(rename = "addDescription")]
    add_description: String,
    /// the skill's level
    level: u8,
    /// how long the skill plays out (in seconds)
    duration: f64,
    /// the skill's modifier
    modifier: SkillModifier,
    /// the chance for the skill to play out (in a ratio from 0 to 1)
    chance: f64,
    /// the skill's radius of effect (in units)
    radius: f64,
}

/// Represents a skill's modifier
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SkillModifier {
    /// the type of modifier
    #[serde(rename = "modifierType")]
    modifier_type: SkillModifierType,
    /// the value of the modifier
    value: f64,
}

/// A list of available skill modifier types.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum SkillModifierType {
    /// the modifier is a percentage
    #[serde(rename = "Percentage")]
    Percentage,
    /// the modifier is either an integer, unsigned integer or a float
    #[serde(rename = "Number")]
    Number,
}